dhall-lsp-server/src/Dhall/LSP/Handlers.hs

```diff
@@ -2,59 +2,346 @@ module Dhall.LSP.Handlers where
+      edits = J.List [J.TextEdit range formatted]
+
+  lspRespond LSP.RspDocumentFormatting request edits
+
+
+executeCommandHandler :: J.ExecuteCommandRequest -> HandlerM ()
+executeCommandHandler request
+  | command == "dhall.server.lint" = executeLintAndFormat request
+  | command == "dhall.server.annotateLet" = executeAnnotateLet request
+  | otherwise = throwE (Warning, "Command '" <> command
+                                   <> "' not known; ignored.")
+  where command = request ^. J.params . J.command
+
+getCommandArguments :: J.FromJSON a => J.ExecuteCommandRequest -> HandlerM a
+getCommandArguments request = do
+  json <- case request ^. J.params . J.arguments of
+    Just (J.List (x : _)) -> return x
+    _ -> throwE (Error, "Failed to execute command; arguments missing.")
+  case J.fromJSON json of
+    J.Success args -> return args
+    _ -> throwE (Error, "Failed to execute command; failed to parse arguments.")
+
+
+-- implements dhall.server.lint
+executeLintAndFormat :: J.ExecuteCommandRequest -> HandlerM ()
+executeLintAndFormat request = do
+  uri <- getCommandArguments request
+  txt <- readUri uri
+
+  (header, expr) <- case parseWithHeader txt of
+    Right res -> return res
+    _ -> throwE (Warning, "Failed to lint dhall code; parse error.")
+
+  let linted = formatExprWithHeader (lint expr) header
+      numLines = Text.length txt
+      range = J.Range (J.Position 0 0) (J.Position numLines 0)
+      edit = J.WorkspaceEdit
+        (Just (HashMap.singleton uri (J.List [J.TextEdit range linted]))) Nothing
+
+  lspRespond LSP.RspExecuteCommand request J.Null
+  lspRequest LSP.ReqApplyWorkspaceEdit J.WorkspaceApplyEdit
+    (J.ApplyWorkspaceEditParams edit)
+
+
+executeAnnotateLet :: J.ExecuteCommandRequest -> HandlerM ()
+executeAnnotateLet request = do
+  args :: J.TextDocumentPositionParams <- getCommandArguments request
+  let uri = args ^. J.textDocument . J.uri
+      line = args ^. J.position . J.line
+      col = args ^. J.position . J.character
+
+  expr <- loadFile uri
+  (welltyped, _) <- case typecheck expr of
+    Left _ -> throwE (Warning, "Failed to annotate let binding; not well-typed.")
+    Right e -> return e
+
+  (Src (SourcePos _ x1 y1) (SourcePos _ x2 y2) _, txt)
+    <- case annotateLet (line, col) welltyped of
+      Right x -> return x
+      Left msg -> throwE (Warning, Text.pack msg)
+
+  let range = J.Range (J.Position (unPos x1 - 1) (unPos y1 - 1))
+                      (J.Position (unPos x2 - 1) (unPos y2 - 1))
+      edit = J.WorkspaceEdit
+        (Just (HashMap.singleton uri (J.List [J.TextEdit range txt]))) Nothing
+
+  lspRequest LSP.ReqApplyWorkspaceEdit J.WorkspaceApplyEdit
+    (J.ApplyWorkspaceEditParams edit)
+
+
+
+
+-- handler that doesn't do anything. Useful for example to make haskell-lsp shut
+-- up about unhandled DidChangeTextDocument notifications (which are already
+-- handled haskell-lsp itself).
+nullHandler :: a -> HandlerM ()
+nullHandler _ = return ()
```

Comment:
Nothing to do here, but it's a shame that `ExceptT` doesn't have a `Monoid` instance of the form:

```haskell
instance (Monad m, Monoid a) => Monoid (ExceptT e m a) where
    mempty = pure mempty

    mappend = liftA2 mappend
```

... because if it did then you could just define:

```haskell
nullHandler = mempty
```

Reply:
Which would save us two characters and thoroughly confuse everybody xD